Local coop

yuripourre opened the pull request:
This is my proposal for local coop. I wanted this feature since I played Diablo for the first time via modem and the game kept disconnecting when someone called.

I am creating this PR so the implementation can be discussed and people can try it. I am really open to suggestions, opinions and even abandon this PR if it's not the direction the community wants to follow. Most of the code is AI generated ~so I don't want to overwhelm anyone asking for code review yet since I am planning to refactor this entire branch.~

I tried to make it completely optional and even if enabled should not be intrusive in any way. The UI is close as possible to the original design, nothing fancy, nothing modern.

### Limitations
- SDL cannot guarantee that the gamepads will be assigned to the same players when the game restarts
- There is no way to control the cursor as a mouse so spells that requires cursor position will not work as expected
- ~No color distinctions (waiting for: #8317)~
- Camera movement is jittery when all players move in the same direction.

```
WARNING: This is a very experimental feature, backups your characters before using them in local coop!
```

## How to activate it
First of all, you should enable the setting "Enable local co-op".

Once in a multiplayer game, the main UI panel is hidden and the local co-op panel is displayed, if the game detects more than 1 gamepad, other players will display the character selection panel. The list of characters is based on the multiplayer saved characters. 

## Gameplay

### Panels
The character panel and inventory are fully functional. I introduced the concept of ownership, whoever opens the panel has control over the panel and can do the same actions as a single player.

### NPCs Interactions
Work similarly to other panels (whoever opens gets ownership).

### Spell Assignment
Alternatively to the default way to assign spells, it's also possible to assign spells to skill slots, so you can have up to 4 spells assigned to your character.

To assign a spell, you have to long press one of the buttons A,B,X or Y. It will open a speedbook and once you press one of those buttons again, the spell/skill is assigned.

Spells can be casted using the assigned button.

### Belt 
I decided to keep the full belt for players. I wanted the local coop panel to be optional for a single player mode in the future.
To use a belt item you must press one of the shoulder buttons.

Left shoulder button highlights items 1 to 4, right shoulder button highlights items 5 to 8. Once the item is highlighted if you press A,B,X,Y the item will be consumed.

<img width="1920" height="1200" alt="Screenshot From 2025-12-12 20-08-34" src="https://github.com/user-attachments/assets/b41726fc-bb3a-4dfb-a036-87b7cde923b1" />

<img width="1920" height="1080" alt="Screenshot From 2025-12-26 15-38-54" src="https://github.com/user-attachments/assets/858a6503-aef1-4c68-9cd5-fd814d571a17" />

https://github.com/user-attachments/assets/0796cbaa-22da-446f-94ed-0c917dba585d


https://github.com/user-attachments/assets/ab5fc544-ff23-4ed1-9213-02e2ae3c29be

Closes #1237

yuripourre commented:
This PR is not perfect:
- Sometimes the players get stuck and cannot move
- The UI is not polished
- I haven't test interactions between players (only monsters)
- I probably broke something during rebase

But I think it's in a good shape to start discussions.

yuripourre commented:
Ready for review

HoofedEar commented:
Incredible work!

My only complaint/issue I've found is when both players are moving in the same direction at the same time, the screen gets a bit jittery? This is at 480p, windowed mode, on Windows 11

https://github.com/user-attachments/assets/f9396063-b4f5-4daf-894a-d81cfa5e88ed

yuripourre commented:
@strich Thanks! You just reminded me that I need to work on the conflicts.

Yeah, I really enjoyed working on this feature. I wanted to do that for a loooong time.

yuripourre commented:
Oh damn! I broke the gamepad logic during the rebase. :sweat_smile:

yuripourre committed:
Add local co-op module

Introduces the local co-op subsystem as a self-contained module under
Source/controls/local_coop/:

- local_coop_constants.hpp  — layout, camera, and input constants
- local_coop_assets.hpp/.cpp — asset loading/unloading for coop HUD
- local_coop_button_mapper.hpp/.cpp — per-player gamepad button routing
- local_coop.hpp/.cpp — core state machine: player join/leave, camera,
  panel layout, input dispatch, hero selection, and visual store wiring

Source/CMakeLists.txt is updated to include the new source files in the
build.

yuripourre committed:
Extend network layer for local co-op

- dvlnet/loopback: message queue now stores sender ID alongside the
  message payload so local co-op players can exchange messages through
  the loopback provider without being collapsed onto player 0
- multi.cpp / init.cpp: call LoadAvailableHeroesForAllLocalCoopPlayers()
  after the save file is read so co-op slots know which heroes are
  eligible
- options.cpp/.h: expose IsLocalCoopEnabled() option

yuripourre committed:
Extend player, save, and lighting systems for local co-op

- player.cpp/.h: track whether a Player slot belongs to a local co-op
  player; add helpers queried by other subsystems
- pfile.cpp/.h: enumerate hero save files for co-op slot selection;
  save/restore all local co-op player heroes on game exit
- loadsave.cpp: load and save extra player slots during level transitions
- portal.cpp: create portals for local co-op players on level change
- lighting.cpp/.h: expose per-player light source so co-op players each
  carry their own light radius independently of player 1
- msg.cpp: route network messages to the correct local co-op player slot

yuripourre committed:
Extend controller and input handling for local co-op

- controller.cpp/controller_motion.cpp: detect hot-plug of additional
  gamepads and assign them to co-op player slots
- devices/game_controller.h, joystick.cpp/.h: expose device index so
  each physical controller maps to a specific co-op player
- game_controls.cpp: dispatch button events to the owning co-op player
  rather than always targeting player 1
- modifier_hints.cpp/.h: show per-player modifier hint overlays
- plrctrls.cpp/.h: route padmenu navigation, attack, skill, belt, and
  inventory actions to the correct local co-op player; extract
  DpadGamepadButtonToAxisDirection helper; use skillSlotSize constant
- touch/event_handlers.cpp: guard touch events against co-op mode

yuripourre committed:
Extend rendering for local co-op

- scrollrt.cpp: split the viewport so each local co-op player sees their
  own screen region; expand the walking tile buffer when any co-op player
  is in motion; apply per-player camera offset; render item/monster/object
  highlights for all co-op players; skip lighting attenuation for local
  players who carry their own light source
- light_render.cpp: apply lighting correctly per player when multiple
  local players are present
- text_render.cpp: clip text to the active player's panel region
- automap.cpp: center the automap on the correct player's position in
  split-screen mode

yuripourre committed:
Extend HUD and UI panels for local co-op

- control.hpp: add ShouldHideMainPanelForLocalCoop() and panel-ownership
  helpers so each co-op player's HUD occupies its own screen quadrant
- control_infobox.cpp: render the info box inside the owning player's
  panel region
- control_panel.cpp: lay out life/mana globes, belt, and skill slots per
  player; draw experience bar; render attribute-available icon
- stores.cpp: gate store access to the interacting player; wire
  InitVisualStore / CloseVisualStore / FreeVisualStoreGFX into the
  town-load and store-close paths via CloseLocalCoopStore
- spells.h: expose speed-spell slot count constant (skillSlotSize)
- cursor.cpp: clamp cursor to the active player's panel region in co-op
- diablo.cpp: call CloseVisualStore from QuestLogKeyPressed and
  DisplaySpellsKeyPressed; initialise and free the visual store on
  town load/unload; route key events to the correct co-op player

yuripourre committed:
Extend gameplay systems for local co-op

- inv.cpp: route inventory open/close, item pick-up, and equipment
  actions to the player whose panel was interacted with
- items.cpp/.h: allow any local co-op player to pick up and use items;
  guard item-use actions by player proximity
- levels/trigs.cpp: trigger level-change portals for all active local
  co-op players when any player steps on a trigger
- missiles.cpp: spawn healing/mana missiles for all local co-op players
  when appropriate (e.g. town potions)
- objects.cpp: allow co-op players to interact with objects inside their
  proximity range
- qol/autopickup.cpp: run auto-pickup logic for each local co-op player
- towners.cpp: allow co-op players to talk to town NPCs
- lua/lua_event.hpp: declare Lua event hooks needed by the co-op module
- utils/hp_mana_units.hpp: extend HP/mana unit helpers used by co-op HUD

yuripourre committed:
Fix compiler warnings unrelated to local co-op

- reencode_dun_cels.cpp: initialise frameSize to 0 to silence
  -Wmaybe-uninitialized on paths where GetReencodedSize returns early
- spell_list.cpp: initialise spells to 0 to silence
  -Wmaybe-uninitialized in IsValidSpeedSpell
- platform/ctr/keyboard.cpp: change eventCount to size_t and introduce
  VkbdEventQueueCapacity constant to silence -Wsign-compare against
  sizeof(events)
